flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/MethodlessRouter.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rest.handler.router;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * This is adopted and simplified code from tv.cntt:netty-router library. For more information check {@link Router}.
+ *
+ * <p>Router that doesn't contain information about HTTP request methods and route
+ * matching orders.
+ */
+final class MethodlessRouter<T> {
+	private static final Logger log = LoggerFactory.getLogger(MethodlessRouter.class);
+
+	// A path pattern can only point to one target
+	private final Map<PathPattern, T> routes = new LinkedHashMap<>();
```

Comment:
Please correct me if I'm wrong but I have a question regarding memory visibility: The thread that updates this map and the Netty event loop threads are different, so there might theoretically be a memory visibility issue if routes are added after the router has been passed to the `RouterHandler`. I don't think that we do this currently, but the API theoretically allows it. I'm wondering whether it makes sense to make the routes immutable, maybe something like creating the routes with a builder:
```java
Routes routes = new RoutesBuilder()
  .addGet(...)
  ...
  .build();
Router router = new Router(routes);
```
Or use a thread-safe map here that also preserves ordering (maybe wrap using `synchronizedMap`).

---
Since we currently rely on correct registration order (e.g. `/jobs/overview` before `/jobs/:id` for correct matching), the immutable approach would allow us to include a utility in `Routes` that sorts pattern as done in `RestServerEndpoint:143`:
```java
Collections.sort(handlers,RestHandlerUrlComparator.INSTANCE);
```

Reply:
Fine with me to not invest more time into this and keep it as is 👍